File: frontend/src/api/adminChallenges.ts
/**
 * Admin-only Team Challenge template management. All routes are gated server-side by the admin
 * allowlist (`require_admin`); the UI only shows the panel when `useAuth().isAdmin` is true.
 *
 * Mirrors the `showdown_bot challenges` CLI: templates are hand-authored content, `rotate` keeps
 * one live instance per category, and `generateInstance` forces one for a single template.
 */

import type { ChallengeCategory, ChallengeGoalType, ChallengeInstance } from './sim';

const API_BASE = import.meta.env.PROD ? '/api' : 'http://127.0.0.1:5000/api';

/** A hand-authored challenge template, with rotation status joined on. */
export type ChallengeTemplate = {
    template_id: string;
    slug: string;
    title: string;
    description: string;
    goal_type: ChallengeGoalType;
    goal_value: { min_wins?: number; target_abbr?: string } | null;
    pts_limit: number | null;
    roster_size: number;
    /** 'any' | comma list of years | 'random_range:lo,hi' */
    year_pool: string;
    /** 'any' | 'worst_record' | comma list of abbrs */
    replaces_pool: string;
    active: boolean;
    player_filters: Record<string, unknown> | null;
    category: ChallengeCategory;
    created_at: string;
    /** Newest instance's created_at, or null if this template has never been instanced. */
    last_instanced_at: string | null;
    /** Unexpired instances pointing at this template (0 or 1 in normal operation). */
    live_instance_count: number;
};

/** The create/edit payload. `min_wins` / `beat_team_abbr` are only read for the matching
 *  `goal_type`; the server re-derives `goal_value` from them. */
export type ChallengeTemplateInput = {
    slug: string;
    title: string;
    description: string;
    goal_type: ChallengeGoalType;
    min_wins?: number | null;
    beat_team_abbr?: string | null;
    category: ChallengeCategory;
    pts_limit?: number | null;
    roster_size: number;
    year_pool: string;
    replaces_pool: string;
    player_filters?: Record<string, unknown> | null;
    active: boolean;
};

export type GeneratedInstance = {
    instance_id: string;
    slug: string;
    year: number;
    replaces_abbr: string;
};

export type RotationResult = {
    pruned: number;
    created: GeneratedInstance[];
    /** Human-readable "category 'x': reason" lines for categories the rotation couldn't fill. */
    skipped: string[];
};

async function fail(res: Response, fallback: string): Promise<never> {
    const body = await res.json().catch(() => ({}));
    throw new Error(body.error || `${fallback} (${res.status})`);
}

function jsonHeaders(token: string) {
    return { 'Content-Type': 'application/json', Authorization: `Bearer ${token}` };
}

export async function fetchAdminChallenges(
    token: string,
): Promise<{ templates: ChallengeTemplate[]; live_instances: ChallengeInstance[] }> {
    const res = await fetch(`${API_BASE}/admin/challenges`, {
        headers: { Authorization: `Bearer ${token}` },
    });
    if (!res.ok) await fail(res, 'Failed to load challenge templates');
    return res.json();
}

export async function createChallengeTemplate(
    token: string,
    input: ChallengeTemplateInput,
): Promise<ChallengeTemplate> {
    const res = await fetch(`${API_BASE}/admin/challenges/templates`, {
        method: 'POST', headers: jsonHeaders(token), body: JSON.stringify(input),
    });
    if (!res.ok) await fail(res, 'Failed to create template');
    return res.json();
}

export async function updateChallengeTemplate(
    token: string,
    templateId: string,
    input: Partial<ChallengeTemplateInput>,
): Promise<ChallengeTemplate> {
    const res = await fetch(`${API_BASE}/admin/challenges/templates/${templateId}`, {
        method: 'PUT', headers: jsonHeaders(token), body: JSON.stringify(input),
    });
    if (!res.ok) await fail(res, 'Failed to update template');
    return res.json();
}

export async function deleteChallengeTemplate(token: string, templateId: string): Promise<void> {
    const res = await fetch(`${API_BASE}/admin/challenges/templates/${templateId}`, {
        method: 'DELETE', headers: { Authorization: `Bearer ${token}` },
    });
    if (!res.ok) await fail(res, 'Failed to delete template');
}

/** Force one live instance for this template now, outside the category rotation. */
export async function generateChallengeInstance(
    token: string,
    templateId: string,
    force = false,
): Promise<GeneratedInstance> {
    const res = await fetch(`${API_BASE}/admin/challenges/templates/${templateId}/instance`, {
        method: 'POST', headers: jsonHeaders(token), body: JSON.stringify({ force }),
    });
    if (!res.ok) await fail(res, 'Failed to generate instance');
    return res.json();
}

/** Prune expired instances, then fill every category that has no live challenge. */
export async function runChallengeRotation(token: string): Promise<RotationResult> {
    const res = await fetch(`${API_BASE}/admin/challenges/rotate`, {
        method: 'POST', headers: jsonHeaders(token),
    });
    if (!res.ok) await fail(res, 'Rotation failed');
    return res.json();
}

File: mlb_showdown_bot/api/admin_challenges.py
"""Admin-only CRUD for Team Challenge templates, plus on-demand instance generation and
rotation. Gated server-side by the admin allowlist (`require_admin`). The frontend surfaces
this in an admin panel above the challenge list; the same operations are available from the
`showdown_bot challenges` CLI.
"""

import traceback

from flask import Blueprint, jsonify, request

from ..core.database.postgres_db import PostgresDB
from ..core.simulation.challenge_generator import (
    MIN_ROSTER_SIZE,
    ChallengeCategory,
    ChallengeError,
    ChallengeGenerator,
    GoalType,
    build_goal_value,
    validate_year_pool,
)
from .user_settings import require_admin

admin_challenges_bp = Blueprint('admin_challenges', __name__)


def _template_fields(payload: dict, *, partial: bool) -> dict:
    """Validate a create (partial=False) or update (partial=True) payload into a column dict
    ready for `create_challenge_template` / `update_challenge_template`. Raises ChallengeError
    with a client-facing message."""
    fields: dict = {}
    present = payload.__contains__

    def want(key: str) -> bool:
        return not partial or present(key)

    if want('slug'):
        slug = (payload.get('slug') or '').strip()
        if not slug:
            raise ChallengeError("slug is required")
        fields['slug'] = slug
    if want('title'):
        title = (payload.get('title') or '').strip()
        if not title:
            raise ChallengeError("title is required")
        fields['title'] = title
    if want('description'):
        fields['description'] = (payload.get('description') or '').strip()

    # goal_type + its goal_value dependents travel together: if any of the three is in the
    # payload, re-derive goal_value so an edit can't leave it stale.
    if present('goal_type') or present('min_wins') or present('beat_team_abbr') or not partial:
        raw_goal = payload.get('goal_type')
        try:
            goal_type = GoalType(raw_goal) if raw_goal is not None else None
        except ValueError:
            raise ChallengeError(f"invalid goal_type '{raw_goal}'")
        if goal_type is None:
            raise ChallengeError("goal_type is required")
        fields['goal_type'] = goal_type.value
        fields['goal_value'] = build_goal_value(goal_type, payload.get('min_wins'), payload.get('beat_team_abbr'))

    if want('category'):
        raw_cat = payload.get('category') or ChallengeCategory.THEMED.value
        try:
            fields['category'] = ChallengeCategory(raw_cat).value
        except ValueError:
            raise ChallengeError(f"invalid category '{raw_cat}'")

    if want('pts_limit'):
        pts = payload.get('pts_limit')
        fields['pts_limit'] = int(pts) if pts not in (None, '') else None
    if want('roster_size'):
        roster = int(payload.get('roster_size') or 25)
        if roster < MIN_ROSTER_SIZE:
            raise ChallengeError(f"roster_size must be at least {MIN_ROSTER_SIZE}")
        fields['roster_size'] = roster

    if want('year_pool'):
        year_pool = (payload.get('year_pool') or 'any').strip()
        validate_year_pool(year_pool)
        fields['year_pool'] = year_pool
    if want('replaces_pool'):
        fields['replaces_pool'] = (payload.get('replaces_pool') or 'any').strip()

    if want('player_filters'):
        pf = payload.get('player_filters')
        if pf in (None, '', {}):
            fields['player_filters'] = None
        elif isinstance(pf, dict):
            fields['player_filters'] = pf
        else:
            raise ChallengeError("player_filters must be a JSON object or null")

    if want('active'):
        fields['active'] = bool(payload.get('active', True))

    return fields


@admin_challenges_bp.route('/admin/challenges', methods=['GET'])
@require_admin
def list_challenges():
    """Every template (with rotation status) plus the instances currently live."""
    try:
        with PostgresDB() as db:
            templates = db.list_challenge_templates()
            live_instances = db.fetch_active_challenges(user_id=None)
        return jsonify({'templates': templates, 'live_instances': live_instances}), 200
    except Exception as exc:
        traceback.print_exc()
        return jsonify({'error': str(exc)}), 500


@admin_challenges_bp.route('/admin/challenges/templates', methods=['POST'])
@require_admin
def create_challenge_template():
    try:
        payload = request.get_json(silent=True) or {}
        fields = _template_fields(payload, partial=False)
        with PostgresDB() as db:
            template_id = db.create_challenge_template(
                slug=fields['slug'], title=fields['title'], description=fields['description'],
                goal_type=fields['goal_type'], goal_value=fields['goal_value'],
                pts_limit=fields.get('pts_limit'), year_pool=fields.get('year_pool', 'any'),
                replaces_pool=fields.get('replaces_pool', 'any'), active=fields.get('active', True),
                player_filters=fields.get('player_filters'), category=fields.get('category', 'themed'),
                roster_size=fields.get('roster_size', 25),
            )
            template = db.get_challenge_template(template_id)
        return jsonify(template), 201
    except ChallengeError as exc:
        return jsonify({'error': str(exc)}), exc.status
    except Exception as exc:
        traceback.print_exc()
        return jsonify({'error': str(exc)}), 500


@admin_challenges_bp.route('/admin/challenges/templates/<template_id>', methods=['PUT'])
@require_admin
def update_challenge_template(template_id: str):
    try:
        payload = request.get_json(silent=True) or {}
        fields = _template_fields(payload, partial=True)
        with PostgresDB() as db:
            template = db.update_challenge_template(template_id, fields)
        if template is None:
            return jsonify({'error': 'Template not found'}), 404
        return jsonify(template), 200
    except ChallengeError as exc:
        return jsonify({'error': str(exc)}), exc.status
    except Exception as exc:
        traceback.print_exc()
        return jsonify({'error': str(exc)}), 500


@admin_challenges_bp.route('/admin/challenges/templates/<template_id>', methods=['DELETE'])
@require_admin
def delete_challenge_template(template_id: str):
    try:
        with PostgresDB() as db:
            result = db.delete_challenge_template(template_id)
        if result == 'not_found':
            return jsonify({'error': 'Template not found'}), 404
        if result == 'in_use':
            return jsonify({'error': 'A live instance still uses this template — deactivate it instead.'}), 409
        return jsonify({'success': True}), 200
    except Exception as exc:
        traceback.print_exc()
        return jsonify({'error': str(exc)}), 500


@admin_challenges_bp.route('/admin/challenges/templates/<template_id>/instance', methods=['POST'])
@require_admin
def generate_challenge_instance(template_id: str):
    """Force one live instance for this template now, outside the category rotation. Body:
    {force?: bool}."""
    try:
        force = bool((request.get_json(silent=True) or {}).get('force'))
        with PostgresDB() as db:
            template = db.get_challenge_template(template_id)
            if template is None:
                return jsonify({'error': 'Template not found'}), 404
            result = ChallengeGenerator(db).instance_from_template(template, force=force)
        return jsonify({
            'instance_id': result.instance_id, 'slug': result.slug,
            'year': result.year, 'replaces_abbr': result.replaces_abbr,
        }), 201
    except ChallengeError as exc:
        return jsonify({'error': str(exc)}), exc.status
    except Exception as exc:
        traceback.print_exc()
        return jsonify({'error': str(exc)}), 500


@admin_challenges_bp.route('/admin/challenges/rotate', methods=['POST'])
@require_admin
def rotate_challenges():
    """Run the weekly rotation on demand: prune old expired instances, then fill every category
    with no live challenge from its least-recently-used active template."""
    try:
        with PostgresDB() as db:
            report = ChallengeGenerator(db).rotate()
        return jsonify({
            'pruned': report.pruned,
            'created': [
                {'instance_id': r.instance_id, 'slug': r.slug, 'year': r.year, 'replaces_abbr': r.replaces_abbr}
                for r in report.created
            ],
            'skipped': report.skipped,
        }), 200
    except Exception as exc:
        traceback.print_exc()
        return jsonify({'error': str(exc)}), 500
